client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java

```diff
@@ -0,0 +1,197 @@
+        requestResponseReplyToAddressTemplate = String.format("%s/%%s/%%s",getName());
+        this.replyToAddress = String.format(
+                requestResponseReplyToAddressTemplate,
+                Objects.requireNonNull(tenantId),
+                UUID.randomUUID());
+
+        final Future<ProtonSender> senderTracker = Future.future();
+        senderTracker.setHandler(r -> {
+            if (r.succeeded()) {
+                LOG.debug("request response client created");
+                this.sender = r.result();
+                creationHandler.handle(Future.succeededFuture((C) this));
+            } else {
+                creationHandler.handle(Future.failedFuture(r.cause()));
+            }
+        });
+
+        final Future<ProtonReceiver> receiverTracker = Future.future();
+        context.runOnContext(create -> {
+            final ProtonReceiver receiver = con.createReceiver(replyToAddress);
+            receiver
+                    .setAutoAccept(true)
+                    .setPrefetch(DEFAULT_SENDER_CREDITS)
+                    .handler((delivery, message) -> {
+                        final Handler<AsyncResult<R>> handler = replyMap.remove(message.getCorrelationId());
+                        if (handler != null) {
+                            R result = getRequestResponseResult(message);
+                            LOG.debug("received response [correlation ID: {}, status: {}]",
+                                    message.getCorrelationId(), result.getStatus());
+                            handler.handle(Future.succeededFuture(result));
+                        } else {
+                            LOG.debug("discarding unexpected response [correlation ID: {}]",
+                                    message.getCorrelationId());
+                        }
+                    }).openHandler(receiverTracker.completer())
+                    .open();
+
+            receiverTracker.compose(openReceiver -> {
+                this.receiver = openReceiver;
+                ProtonSender sender = con.createSender(String.format(requestResponseAddressTemplate, tenantId));
+                sender
+                        .setQoS(ProtonQoS.AT_LEAST_ONCE)
+                        .openHandler(senderTracker.completer())
+                        .open();
+            }, senderTracker);
+        });
+    }
+
+    /**
+     * Build a Proton message with a provided subject (serving as the operation that shall be invoked).
+     * The message can be extended by arbitrary application properties passed in.
+     *
+     * @param subject The subject system property of the message.
+     * @param appProperties The map containing arbitrary application properties.
+     *                      Maybe null if no application properties are needed.
+     * @return The Proton message constructed from the provided parameters.
+     */
+    protected Message createMessage(final String subject, final Map<String, Object> appProperties) {
```

Comment:
the passed in `appProperties` should be checked/filtered for compatible value types as we do in `AbstractSender.setApplicationProperties`. Maybe we can pull up that method into `AbstractHonoClient` and use it here as well?